### ISSUE TITLE
PyTorch fails to execute BatchNorm2d, Conv2d, and Conv3d on AMD Ryzen AI Max+ 395 (gfx1151) with ROCm 7.1

### ISSUE THREAD

**Issue by wang-xiri** [ROCmSoftwarePlatform/pytorch#2837]
**Description:**

I am attempting to run standard PyTorch operations—including `torch.nn.BatchNorm2d`, `torch.nn.Conv2d`, and `torch.nn.Conv3d`—on an AMD Ryzen AI Max+ 395 integrated GPU (GFX architecture: `gfx1151`) using ROCm 7.1.

I have tested both:
- A native installation of ROCm 7.1 with PyTorch 2.8.0 (built for ROCm)
- The official Docker image: `rocm/pytorch:rocm7.1_ubuntu24.04_py3.12_pytorch_release_2.8.0`

In both environments, attempting to execute these operations results in failures with MIOpen-related errors rather than silently falling back to CPU execution. Specifically, PyTorch correctly detects the GPU (torch.cuda.is_available() returns True, and torch.cuda.get_device_name(0) reports "AMD Radeon Graphics"), but when attempting to run BatchNorm2d, Conv2d, or Conv3d layers, the operations fail with errors indicating that these functions are not supported on the current GPU architecture (gfx1151). Detailed error logs will be attached separately.

Given that `gfx1151` is a newly released RDNA 3.5 architecture, I suspect missing MIOpen kernel support for this target in ROCm 7.1. However, clear error messaging or functional fallback behavior would be expected.

Please consider this a request to either:
1. Add proper MIOpen kernel support for `gfx1151`, or  
2. Improve error reporting when unsupported architectures are used.

**Here is my python check script**

``` python
import torch
import os
os.environ['MIOPEN_LOG_LEVEL'] = '7'
os.environ['MIOPEN_BN_DISABLE_INLINE_ASM'] = '1'

os.environ['MIOPEN_DISABLE_CACHE'] = '1'
os.environ['MIOPEN_FIND_MODE'] = 'NO_DB'

def check_BatchNorm2d_half():
    try:
        print('check BatchNorm2d_half')
        bn = torch.nn.BatchNorm2d(64).cuda().half()
        x = torch.randn(2, 64, 32, 32).cuda().half()
        y = bn(x)
        print("OK")
    except Exception as e:
        print("\n--- ERROR ---")
        print(f"Failed as expected: {e}")

def check_Conv2d():
    try:
        print('check Conv2d')
        conv = torch.nn.Conv2d(
            in_channels=64,
            out_channels=64,
            kernel_size=3,
            padding=1
        ).cuda()

        input_tensor = torch.randn(
            1, 64, 16, 32
        ).cuda()

        y = conv(input_tensor)

        print("OK")
    except Exception as e:
        print("\n--- ERROR ---")
        print(f"Failed as expected: {e}")

def check_Conv3d_half():
    try:
        print('check Conv3d_half')
        conv = torch.nn.Conv3d(
            in_channels=64,
            out_channels=64,
            kernel_size=3,
            padding=1
        ).cuda().half()

        input_tensor = torch.randn(
            1, 64, 16, 32, 32
        ).cuda().half()

        y = conv(input_tensor)

        print("Attempting to perform 3D convolution with fp16...")
    except Exception as e:
        print("\n--- ERROR ---")
        print(f"Failed as expected: {e}")
def check_Conv3d():
    # Check if ROCm device is available
    if not torch.cuda.is_available():
        print("ROCm device not found.")
        exit()
    
    print("Checking Conv3d...")

    print(f"Device: {torch.cuda.get_device_name(0)}")

    # Example parameters for a 3D convolution
    # These may need to be adjusted to match the exact failing case
    try:
        conv_layer = torch.nn.Conv3d(
            in_channels=320,
            out_channels=320,
            kernel_size=3,
            padding=1
        ).cuda().half()

        input_tensor = torch.randn(
            1, 320, 16, 32, 32
        ).cuda().half()

        print("Attempting to perform 3D convolution with fp16...")
        output = conv_layer(input_tensor)
        print("Success! The convolution worked.")

    except Exception as e:
        print("\n--- ERROR ---")
        print(f"Failed as expected: {e}")
def check_BatchNorm2d():    
    try:
        print('check BatchNorm2d')
        bn = torch.nn.BatchNorm2d(64).cuda()
        x = torch.randn(2, 64, 32, 32).cuda()
        y = bn(x)
        print("OK")
    except Exception as e:
        print("\n--- ERROR ---")
        print(f"Failed as expected: {e}")



def check_cuda():
    try: 

        # 检测设备
        if torch.cuda.is_available():
            device = torch.device("cuda")
            print("\n🟢 使用 NVIDIA CUDA 设备")
            print("CUDA 设备数量:", torch.cuda.device_count())
            print("当前 CUDA 设备索引:", torch.cuda.current_device())
            print("当前 CUDA 设备名称:", torch.cuda.get_device_name(0))
    except Exception as e:
        print("\n--- ERROR ---")
        print(f"Failed as expected: {e}")
def check_hip():
    try:
        if torch.version.hip is not None:
            
            device = torch.device("cuda")  # 注意：ROCm 也使用 "cuda" 后端！
            print("\n🟢 使用 AMD ROCm 设备")
            print("ROCm 版本:", torch.version.hip)
            # ROCm 下同样用 torch.cuda.* 查询 GPU
            if torch.cuda.is_available():  # 在 ROCm 环境中，这个也会返回 True
                print("ROCm GPU 数量:", torch.cuda.device_count())
                print("GPU 名称:", torch.cuda.get_device_name(0))
                print(f"HIP version: {torch.version.hip}")
                print(f"ROCm is available with {torch.cuda.device_count()} GPU(s)")
                print(f"Current device: {torch.cuda.current_device()}")
                print(f"Device name: {torch.cuda.get_device_name(0)}")
            else:
                print("⚠️ ROCm 已安装但未检测到可用 GPU")
        else:
            print("⚠️ ROCm 未安装或未检测到可用 GPU")
    except Exception as e:
        print("\n--- ERROR ---")
        print(f"Failed as expected: {e}")

def check_matmul():
    try:
        # AMD ROCm 测试
        print("\n开始测试矩阵乘法...cuda")
        device = "cuda"
        x = torch.randn(1024, 1024).to(device)
        y = torch.randn(1024, 1024).to(device)
        z = torch.matmul(x, y)
        print("\n✅ cuda 矩阵乘法成功完成")
        # 清空缓存
        torch.cuda.empty_cache()
    except Exception as e:
        print("\n--- ERROR ---")
        print(f"Failed as expected: {e}")

def check_matmul_cuda_0():
    try:

        print("\n开始测试矩阵乘法...cuda:0")
        device = "cuda:0"
        x = torch.randn(1024, 1024).to(device)
        y = torch.randn(1024, 1024).to(device)
        z = torch.matmul(x, y)
        print("\n✅ cuda:0 矩阵乘法成功完成")
        # 清空缓存
        torch.cuda.empty_cache()

            # AMD ROCm 测试
        pass
    except Exception as e:
        print("\n--- ERROR ---")
        print(f"Failed as expected: {e}")


if __name__ == "__main__":
    check_cuda()
    check_hip()
    check_matmul()
    check_matmul_cuda_0()
    check_BatchNorm2d_half()
    print("\n********************************开始测试Conv3d_half...")
    check_Conv3d_half()
    print("\n********************************开始测试Conv3d...")
    check_Conv3d()
    print("\n********************************开始测试Conv2d...")
    check_Conv2d()
    print("\n********************************开始测试BatchNorm2d...")
    check_BatchNorm2d()

```

**And here is the output:**

```
🟢 使用 NVIDIA CUDA 设备
CUDA 设备数量: 1
当前 CUDA 设备索引: 0
当前 CUDA 设备名称: Radeon 8060S Graphics

🟢 使用 AMD ROCm 设备
ROCm 版本: 7.1.25413-7721681424
ROCm GPU 数量: 1
GPU 名称: Radeon 8060S Graphics
HIP version: 7.1.25413-7721681424
ROCm is available with 1 GPU(s)
Current device: 0
Device name: Radeon 8060S Graphics

开始测试矩阵乘法...cuda

✅ cuda 矩阵乘法成功完成

开始测试矩阵乘法...cuda:0

✅ cuda:0 矩阵乘法成功完成
check BatchNorm2d_half
OK

********************************开始测试Conv3d_half...
check Conv3d_half
MIOpen(HIP): Info [get_device_name] Raw device name: gfx1151
MIOpen(HIP): Info [Handle] stream: 0, device_id: 0
MIOpen(HIP): Info [get_device_name] Raw device name: gfx1151
MIOpen(HIP): Info [SetStream] stream: 0, device_id: 0
MIOpen Error: stoul

--- ERROR ---
Failed as expected: miopenStatusUnknownError

********************************开始测试Conv3d...
Checking Conv3d...
Device: Radeon 8060S Graphics
Attempting to perform 3D convolution with fp16...
MIOpen(HIP): Info [get_device_name] Raw device name: gfx1151
MIOpen(HIP): Info [SetStream] stream: 0, device_id: 0
MIOpen Error: stoul

--- ERROR ---
Failed as expected: miopenStatusUnknownError
check BatchNorm2d
MIOpen(HIP): Info [get_device_name] Raw device name: gfx1151
MIOpen(HIP): Info [SetStream] stream: 0, device_id: 0
MIOpen(HIP): Info [AmdRocmMetadataVersionDetect] ROCm MD version AMDHSA_COv3, HIP version 7.1.25413, MIOpen version 3.5.1.10cc739c23
MIOpen(HIP): Info2 [GetInvoker] Returning an invoker for problem 64x1x32x32x2xNCHWxNCHWxFP32xFP32xFP32xFP32xFP32xFP32x0xTrnx1x1x1x0 and algorithm miopenBatchNormForwardTrainingSpatial
MIOpen(HIP): Info2 [GetFound1_0] No invokers found for 64x1x32x32x2xNCHWxNCHWxFP32xFP32xFP32xFP32xFP32xFP32x0xTrnx1x1x1x0
MIOpen(HIP): Info [FindSolutionImpl] BnFwdTrainingSpatial
MIOpen(HIP): Info2 [GetLibPath] Lib Path: "/home/zev/miniconda3/lib/python3.13/site-packages/_rocm_sdk_libraries_gfx1151/lib/libMIOpen.so.1"
MIOpen(HIP): Info2 [GetPerfDbPathFile] inexact perf database search
MIOpen(HIP): Info2 [GetPerfDbPathFile] Iterating over perf db directory "/home/zev/miniconda3/lib/python3.13/site-packages/_rocm_sdk_libraries_gfx1151/share/miopen/db"
MIOpen(HIP): Info [IsNetworkedFilesystem] Filesystem type at '"/home/zev/.config/miopen/"' is: 0xef53 'EXT2/3/4_SUPER_MAGIC'
MIOpen(HIP): Info [Measure] ReadonlyRamDb::Prefetch time: 5e-05 ms
MIOpen(HIP): Info [Prefetch] File is unreadable: "/home/zev/.config/miopen/batchnorm_gfx1151_20.HIP.3_5_1_10cc739c23.udb.txt"
MIOpen(HIP): Info [Measure] RamDb::Prefetch time: 0.128463 ms
MIOpen(HIP): Info2 [FindRecordUnsafe] Looking for key 2x2x64x32x32x1x1x1x0xNCHWxTrnxFP32x1 in cache for file "/home/zev/.config/miopen/batchnorm_gfx1151_20.HIP.3_5_1_10cc739c23.udb.txt"
MIOpen(HIP): Info2 [FindRecord] Looking for key 2x2x64x32x32x1x1x1x0xNCHWxTrnxFP32x1 in file ""
MIOpen(HIP): Info2 [Measure] Db::Load time: 0.011933 ms
MIOpen(HIP): Info [FindSolutionImpl] Perf Db: record not found for: BnFwdTrainingSpatial
MIOpen(HIP): Info [GetDefaultPerformanceConfig] Variant3-1
MIOpen(HIP): Info2 [SearchForSolutions] BnFwdTrainingSpatial: Success.
MIOpen(HIP): Info2 [PrepareInvoker] Preparing kernel: MIOpenBatchNormFwdTrainSpatial
MIOpen(HIP): Info [PrintVersionImpl] COMgr v.3.0.0
MIOpen(HIP): Info2 [SetIsaName] amdgcn-amd-amdhsa--gfx1151
<inline asm>:14:20: error: not a valid operand.
v_add_f32 v9 v9 v9 row_bcast:15 row_mask:0xa
                   ^
<inline asm>:15:20: error: not a valid operand.
v_add_f32 v8 v8 v8 row_bcast:15 row_mask:0xa
                   ^
<inline asm>:17:20: error: not a valid operand.
v_add_f32 v9 v9 v9 row_bcast:31 row_mask:0xc
                   ^
<inline asm>:18:20: error: not a valid operand.
v_add_f32 v8 v8 v8 row_bcast:31 row_mask:0xc
                   ^
MIOpen(HIP): Error [Do] 'amd_comgr_do_action(kind, handle, in.GetHandle(), out.GetHandle())' AMD_COMGR_ACTION_CODEGEN_BC_TO_RELOCATABLE: ERROR (1)
MIOpen(HIP): Error [BuildOcl] comgr status = ERROR (1)
MIOpen(HIP): Warning [BuildOcl] error: cannot compile inline asm
error: cannot compile inline asm
error: cannot compile inline asm
error: cannot compile inline asm
4 errors generated.

MIOpen Error: /therock/src/rocm-libraries/projects/miopen/src/hipoc/hipoc_program.cpp:299: Code object build failed. Source: MIOpenBatchNormFwdTrainSpatial.cl

--- ERROR ---
Failed as expected: miopenStatusUnknownError
double free or corruption (!prev)
已中止 (核心已转储)
```
    

### Versions

```
Collecting environment information...
PyTorch version: 2.8.0+rocm7.1.0.git7a520360
Is debug build: False
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: 7.1.25424-4179531dcd

OS: Ubuntu 24.04.3 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.39

Python version: 3.12.3 (main, Aug 14 2025, 17:47:21) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.14.0-36-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: AMD Radeon Graphics (gfx1151)
Nvidia driver version: Could not collect
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: 7.1.25424
MIOpen runtime version: 3.5.1
Is XNNPACK available: True

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           48 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  32
On-line CPU(s) list:                     0-31
Vendor ID:                               AuthenticAMD
Model name:                              AMD RYZEN AI MAX+ 395 w/ Radeon 8060S
CPU family:                              26
Model:                                   112
Thread(s) per core:                      2
Core(s) per socket:                      16
Socket(s):                               1
Stepping:                                0
Frequency boost:                         enabled
CPU(s) scaling MHz:                      51%
CPU max MHz:                             5187.0000
CPU min MHz:                             599.0000
BogoMIPS:                                5988.67
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good amd_lbr_v2 nopl xtopology nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba perfmon_v2 ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local user_shstk avx_vnni avx512_bf16 clzero irperf xsaveerptr rdpru wbnoinvd cppc arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif x2avic v_spec_ctrl vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid bus_lock_detect movdiri movdir64b overflow_recov succor smca fsrm avx512_vp2intersect flush_l1d amd_lbr_pmc_freeze
Virtualization:                          AMD-V
L1d cache:                               768 KiB (16 instances)
L1i cache:                               512 KiB (16 instances)
L2 cache:                                16 MiB (16 instances)
L3 cache:                                64 MiB (2 instances)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-31
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Mitigation; IBPB on VMEXIT only
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; STIBP always-on; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Mitigation; IBPB on VMEXIT

Versions of relevant libraries:
[pip3] numpy==2.3.4
[pip3] torch==2.8.0+rocm7.1.0.lw.git7a520360
[pip3] torchaudio==2.8.0+rocm7.1.0.git6e1c7fe9
[pip3] torchvision==0.23.0+rocm7.1.0.git824e8c87
[pip3] triton==3.4.0+rocm7.1.0.gitf9e5bf54
[conda] Could not collect
```

**Comment by wang-xiri** [comment]
reinstall rocm 7.11.0a20251208, this bug is fixed.
